src/Microsoft.Health.Fhir.Core.UnitTests/Features/Operations/Export/ExportJobTaskTests.cs

```diff
@@ -301,6 +301,73 @@ public async Task GivenDeleteSecretFailed_WhenExecuted_ThenJobStatusShouldBeUpda
         }
 
+        [Fact]
+        public async Task GivenAnExportJobToResume_WhenExecuted_ThenItShouldExportAllRecordsAsExpected()
+        {
+            // We are using the SearchService to throw an exception in order to simulate the export job task
+            // "crashing" while in the middle of the process.
+            _exportJobConfiguration.NumberOfPagesPerCommit = 2;
+
+            int numberOfCalls = 0;
+            int numberOfSuccessfulPages = 2;
+
+            _searchService.SearchAsync(
+                Arg.Any<string>(),
+                Arg.Any<IReadOnlyList<Tuple<string, string>>>(),
+                _cancellationToken)
+                .Returns(x =>
+                {
+                    int count = numberOfCalls++;
+
+                    if (count == numberOfSuccessfulPages)
+                    {
+                        throw new Exception();
+                    }
+
+                    return CreateSearchResult(
+                        new[]
+                        {
+                            new ResourceWrapper(
+                                count.ToString(CultureInfo.InvariantCulture),
+                                "1",
+                                "Patient",
+                                new RawResource("data", Core.Models.FhirResourceFormat.Json),
+                                null,
+                                DateTimeOffset.MinValue,
+                                false,
+                                null,
+                                null,
+                                null),
+                        },
+                        continuationToken: "ct");
+                });
+
+            await _exportJobTask.ExecuteAsync(_exportJobRecord, _weakETag, _cancellationToken);
+
+            string exportedIds = _inMemoryDestinationClient.GetExportedData(new Uri("Patient.ndjson", UriKind.Relative));
+
+            Assert.Equal("01", exportedIds);
+            Assert.NotNull(_exportJobRecord.Progress);
+
+            // We create a new export job task here to simulate the worker picking up the "old" export job record
+            // and resuming the export process. The export destination client contains data that has
+            // been committed up until the "crash".
+            var secondExportJobTask = new ExportJobTask(
+                _fhirOperationDataStore,
+                _secretStore,
+                Options.Create(_exportJobConfiguration),
+                _searchService,
+                _resourceToByteArraySerializer,
+                _exportDestinationClientFactory,
+                NullLogger<ExportJobTask>.Instance);
+
+            numberOfSuccessfulPages = 6;
+            await secondExportJobTask.ExecuteAsync(_exportJobRecord, _weakETag, _cancellationToken);
+
+            exportedIds = _inMemoryDestinationClient.GetExportedData(new Uri("Patient.ndjson", UriKind.Relative));
```

Comment:
In the real world scenario, the destination client will be re-created so perhaps instead of reuse the in memory destination client, reinstantiate it and only capture the new entires?

also, shouldn't it also capture 5?

Reply:
The reason for the discrepancy between the `numberOfSuccessfulPages` and `exportedIds` is because the `numberOfCalls` is incremented before the exception is thrown in the search service. That's why 2 ends up being skipped and we only end up with 0134 (and no 2 or 5). I'll update the code to increment `numberOfCalls` after the exception is thrown so that the resulting numbers are in order.